mango/tuner.py

```diff
@@ -60,6 +67,14 @@ def strategy_is_penalty(self):
         def strategy_is_clustering(self):
             return self.parallel_strategy == 'clustering'
 
+        def early_stop(self, results):
+            if self.early_stopping is None:
+                return False
+
+            results = copy.deepcopy(results)
```

Comment:
is this expensive?

Reply:
Good point. Based on some benchmarks: it takes 1e-5 secs per optimizer iteration to do `deepcopy` on the results data structure. This orders of magnitude lower than the Bayesian optimizer overhead.

Benchmark code:

```
import numpy as np
import copy
import timeit
from scipy.stats import uniform, loguniform
from sklearn.model_selection import ParameterSampler

space = {
        'f': range(10),
        'h': uniform(-10, 20),
        'e': ['cat1', 1, 'cat2'],
        'c': [1, 2, 3],
        'a': ['const'],
        'g': loguniform(0.001, 100),
        'b': [10],
        'd': uniform(0, 1),
        'i': [True, False]
    }


iters = 1000
results = dict(
    objective_values=np.random.uniform(iters),
    params_tries=list(ParameterSampler(space, n_iter=iters)),
    surrogate_values=np.random.uniform(iters),
    best_objective=np.random.uniform(),
    best_params=list(ParameterSampler(space, n_iter=1))[0],
)


def run():
    res = copy.deepcopy(results)


n = 1000
print(timeit.timeit(run, number=n) / n)
```

Reply:
nice!